portal/src/helpers/domain_parsing.ts

```diff
@@ -28,22 +19,41 @@ export function getDomain(url: URL): string {
 * @returns Path object e.g. { subdomain: "subname.name", path: "/index.html"}
 */
 export function getSubdomainAndPath(url: URL): Path | null {
+    return splitUrl(url).path;
+}
+
+/**
+* Given a URL, returns the extracted parts of it.
+* @param url e.g. "https://subname.name.walrus.site/"
+* @returns Path object e.g. { domain: name.walrus.site, {subdomain: "subname", path: "/index.html"}}
+*/
+function splitUrl(url: URL): UrlExtract {
     const parsed = parseDomain(url.hostname);
-    const path = url.pathname == "/" ? "/index.html" : removeLastSlash(url.pathname)
+    let domain: string | null = null;
+    let subdomain: string | null = null;
     if (parsed.type === ParseResultType.Listed) {
+        domain = domain = parsed.domain + "." + parsed.topLevelDomains.join(".")
+        subdomain = parsed.subDomains.join(".")
+    } else if (parsed.type === ParseResultType.Reserved) {
+        domain = parsed.labels[parsed.labels.length - 1];
+        subdomain = parsed.labels.slice(0, parsed.labels.length - 1).join('.');
+    } else {
         return {
-            subdomain: parsed.subDomains.join("."),
-            path
-        } as Path;
-    } else if ( parsed.type === ParseResultType.Reserved) {
-        return {
-            subdomain: parsed.labels.slice(0, parsed.labels.length-1).join('.'),
-            path
-        } as Path;
+            domain: null,
+            path: null
+        }
     }
-    return null;
+
+    return {
+        domain,
+        path: {
```

Comment:
Nit: having `path` as a name nested twice is a bit odd... I think I made a mistake initially in calling the class `Path` (it was for something else at the beginning, and the name stuck). Can we have the `UrlExtract` struct be flat, and then have an instance method that returns the `Path` struct? (also maybe rename `Path`).

Reply:
I wouldn't disagree with this, I will rename the `Path` to something else, I had this in mind too.
 
On the other hand, since `UrlExtract` is a type definition and not a struct, it's not possible to have getter methods in it to fetch `Path` from the flattened `UrlExtract`. I think I will keep `UrlExtract` as is for now, bot with a new name for path.